storage/src/object_store/ecs_s3_client.hpp

```diff
@@ -0,0 +1,412 @@
+                                     << ", msg: " << rData.errorMessage);
+      if (rData.status == S3Status::S3StatusHttpErrorNotFound || rData.status == S3Status::S3StatusErrorNoSuchBucket ||
+          rData.status == S3Status::S3StatusErrorNoSuchKey)
+        return Status::NotFound("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+
+      return Status::GeneralError("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+    }
+  }
+
+  Status test_bucket_internal() {
+    assert(init_);
+    if (should_net_fail()) return Status::GeneralError("Network failure simulated");
+    ResponseData rData;
+    S3ResponseHandler rHandler;
+    rHandler.completeCallback = responseCompleteCallback;
+    rHandler.propertiesCallback = propertiesCallback;
+    char location[100];
+
+    S3_test_bucket(context_.protocol,
+                   context_.uriStyle,
+                   context_.accessKeyId,
+                   context_.secretAccessKey,
+                   context_.hostName,
+                   context_.bucketName,
+                   100,
+                   location,
+                   NULL,
+                   &rHandler,
+                   &rData);
+    LOG_DEBUG(logger_,
+              "test_bucket bucket: " << context_.bucketName << ", status: " << rData.status
+                                     << ", msg: " << rData.errorMessage);
+    if (rData.status == S3Status::S3StatusOK)
+      return Status::OK();
+    else {
+      LOG_ERROR(logger_,
+                "test_bucket bucket: " << context_.bucketName << ", status: " << rData.status
+                                       << ", msg: " << rData.errorMessage);
+      if (rData.status == S3Status::S3StatusHttpErrorNotFound || rData.status == S3Status::S3StatusErrorNoSuchBucket)
+        return Status::NotFound("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+
+      return Status::GeneralError("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+    }
+  }
+
+  struct ResponseData {
+    S3Status status = S3Status::S3StatusOK;
+    string errorMessage;
+  };
+
+  /**
+   * @brief Represents intermidiate data for get operation and is passed to
+   * the lambda callback.
+   *
+   */
+  struct GetObjectResponseData : public ResponseData {
+    GetObjectResponseData(size_t linitialLength) : data(new char[linitialLength]), dataLength(linitialLength) {}
+
+    void CheckAndResize(int& nextReadLength) {
+      if (readLength + nextReadLength <= dataLength) return;
+
+      dataLength *= 2;
+      char* newData = new char[dataLength];
+      memcpy(newData, data, readLength);
+      delete[] data;
+      data = newData;
+    }
+
+    void Write(const char* _data, int& dataSize) {
+      CheckAndResize(dataSize);
+      memcpy(data + readLength, _data, dataSize);
+      readLength += dataSize;
+    }
+
+    ~GetObjectResponseData() { delete[] data; }
+
+    char* data = nullptr;
```

Comment:
Can we please use vectors for data storage instead of raw pointers to make this exception safe and prevent leaks, etc...